pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c

```diff
@@ -0,0 +1,145 @@
+    g_perror("mount");
+}
+
+void bind_mount_host(const gchar *host, const gchar *guest) {
+  g_autofree gchar *mountpoint = g_build_filename(guest, "host", NULL);
+  bind_mount(host, mountpoint);
+}
+
+void bind_mount_item(const gchar *host, const gchar *guest, const gchar *name) {
+  g_autofree gchar *source = g_build_filename(host, name, NULL);
+  g_autofree gchar *target = g_build_filename(guest, name, NULL);
+
+  if (G_LIKELY(g_file_test(source, G_FILE_TEST_IS_DIR)))
+    bind_mount(source, target);
+}
+
+void bind(const gchar *host, const gchar *guest) {
+  g_autoptr(GError) err = NULL;
+  g_autoptr(GDir) dir = g_dir_open(host, 0, &err);
+
+  if (err != NULL)
+    g_perror("g_dir_open");
+
+  const gchar *item;
+
+  while (item = g_dir_read_name(dir))
+    if (!g_strv_contains(bind_blacklist, item))
+      bind_mount_item(host, guest, item);
+
+  bind_mount_host(host, guest);
+}
+
+void spit(const char *path, char *fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+
+  FILE *f = g_fopen(path, "w");
+
+  if (f == NULL)
+    perror("g_fopen");
+
+  if (g_vfprintf(f, fmt, args) < 0)
+    perror("g_vfprintf");
+
+  if (fclose(f) < 0)
+    perror("fclose");
+}
+
+int nftw_remove(const char *path, const struct stat *sb, int type,
+                struct FTW *ftw) {
+  return remove(path);
+}
+
+int main(gint argc, gchar **argv) {
+  const gchar *self = *argv++;
+
+  if (argc < 2) {
+    g_printerr("Usage: %s command [arguments...]\n", self);
+    return 1;
+  }
+
+  if (g_getenv("NIX_CHROOTENV"))
+    g_warning("chrootenv doesn't stack!");
+
+  g_setenv("NIX_CHROOTENV", "", FALSE);
+
+  g_autofree gchar *prefix =
+      g_build_filename(g_get_tmp_dir(), "chrootenvXXXXXX", NULL);
+
+  if (g_mkdtemp_full(prefix, 0755) < 0)
+    g_perror("g_mkdtemp_full");
+
+  pid_t cpid = fork();
+
+  if (cpid < 0)
+    g_perror("fork");
+
+  else if (cpid == 0) {
+    uid_t uid = getuid();
+    gid_t gid = getgid();
+
+    if (unshare(CLONE_NEWNS | CLONE_NEWUSER) < 0) {
+      g_message("Requires Linux version >= 3.19 built with CONFIG_USER_NS");
+      if (g_file_test("/proc/sys/kernel/unprivileged_userns_clone",
+                      G_FILE_TEST_EXISTS))
+        g_message("Run: sudo sysctl -w kernel.unprivileged_userns_clone=1");
+
+      g_perror("unshare");
+    }
+
+    spit("/proc/self/setgroups", "deny");
+    spit("/proc/self/uid_map", "%d %d 1", uid, uid);
+    spit("/proc/self/gid_map", "%d %d 1", gid, gid);
+
+    bind("/", prefix);
+
+    if (chroot(prefix) < 0)
+      g_perror("chroot");
+
+    if (execvp(*argv, argv) < 0)
+      g_perror("execvp");
+  }
+
+  else {
+    int status;
+
+    if (waitpid(cpid, &status, 0) < 0)
+      g_perror("waitpid");
+
+    if (nftw(prefix, nftw_remove, getdtablesize(),
+             FTW_DEPTH | FTW_MOUNT | FTW_PHYS) < 0)
+      g_perror("nftw");
+
+    if (WIFEXITED(status))
+      return WEXITSTATUS(status);
+
+    else if (WIFSIGNALED(status))
+      kill(getpid(), WTERMSIG(status));
+  }
```

Comment:
It is not obvious that the equivalent of `return EX_OSERR;` of the old version is unneeded in the new version: the signal that terminated the child process may be ignored by default and not terminate chrootenv.

Reply:
If signal is ignored, it will return 0. This seems to be logical behavior.

Reply:
If the child overrode the default of ignoring a signal and died on receiving it, it did not exit cleanly and chrootenv should not return 0.

Reply:
You are right, it's better to return 1 in this case. I've added it, see https://github.com/NixOS/nixpkgs/pull/33139/commits/ac40a57cea695493e0da2b91530688ea2da4f240.